PR9

AlexandrSid opened the pull request:
Exceptiono handler added

AlexandrSid committed:
url shortener hash provider

AlexandrSid committed:
url shortener hash repository

AlexandrSid committed:
url shortener async hash refilling

AlexandrSid committed:
url shortener.
HashCache added with scheduled refill of local cache

AlexandrSid committed:
url shortener.
Endpoint to process long url added

AlexandrSid committed:
url shortener.
Endpoint for redirect to original

init refactoring

AlexandrSid committed:
url shortener.
Endpoint for redirect to original

Redirect to original URL endpoint added tested

AlexandrSid committed:
url shortener.
Expired clean up

AlexandrSid committed:
url shortener.
refactor

AlexandrSid committed:
url shortener.
exception handler

AlexandrSid committed:
url shortener.
Refactoring, made /url endpoint required x-user-id header when /{hash} is not

miked4king commented on src/main/java/faang/school/urlshortenerservice/Base62Encoder.java:
```diff
@@ -0,0 +1,40 @@
+package faang.school.urlshortenerservice;
```
давай оставим в корне проекта только UrlShortnerApplication класс

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,30 @@
+@RestController
+@RequiredArgsConstructor
+public class UrlController {
+    private final UrlValidator validator = new UrlValidator(new String[]{"http", "https"});
```
new UrlValidator(new String[]{"http", "https"}) - давай отдельный бин создадим

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,30 @@
+import org.apache.commons.validator.routines.UrlValidator;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.servlet.view.RedirectView;
+
```
@RequestMapping(/api/v1) /api/v1 - из конфига

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,30 @@
+    public String uploadLong(@RequestBody LongUrlDto longUrlDto) {
+        String longUrl = longUrlDto.getLongUrl();
+        if (!validator.isValid(longUrl)) {
+            throw new IllegalArgumentException("provided url is not valid");
```
предложил бы добавить пример валидного url в текст с ошибкой

miked4king commented on src/main/java/faang/school/urlshortenerservice/exception/UrlExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+package faang.school.urlshortenerservice.exception;
```
лучше перенести в пакет с контроллером, в пакете exception сами исключения содержатся

miked4king commented on src/main/java/faang/school/urlshortenerservice/HashCache.java:
```diff
@@ -0,0 +1,48 @@
+    private void calculateCacheGenThreshold() {
+        cacheGenThreshold =
+                constantsProperties.getLocalCachingSize() * constantsProperties.getGenerationThresholdPercent() / 100;
+    }
```
расположить метод после полей класса

miked4king commented on src/main/java/faang/school/urlshortenerservice/HashCache.java:
```diff
@@ -0,0 +1,48 @@
+
+@Component
+@RequiredArgsConstructor
+public class HashCache {
```
добавить интерфейс

miked4king commented on src/main/java/faang/school/urlshortenerservice/HashCache.java:
```diff
@@ -0,0 +1,48 @@
+        return cache.poll();
+    }
+
+    @Scheduled(cron = "${spring.scheduling.hash_check_refill}")
```
давай переработаем без планировщика
при запуске в postconstruct проинициализируем cache, при вызове getHash при необходимости дозаполним

miked4king commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,55 @@
+@Slf4j
+@Repository
+@RequiredArgsConstructor
+public class HashRepository {
```
интерфейсы для классов в данном пакете добавить

miked4king commented on src/main/java/faang/school/urlshortenerservice/Base62Encoder.java:
```diff
@@ -0,0 +1,40 @@
+        return (number * properties.getShuffleKey()) % (properties.getMaxNumber() + 1);
+    }
+
+    private String encodeToBase62(long shuffledNumber) {
```
на твое усмотрение (я бы использовал готовую либу)

AlexandrSid replied:
https://github.com/sqids/sqids-java - новый провайдер

miked4king commented on src/main/java/faang/school/urlshortenerservice/HashGenerator.java:
```diff
@@ -0,0 +1,35 @@
+        if (hashRepository.countHashes() > generatorThreshold) return;
+        List<Long> uniqueNumbers = hashRepository.getUniqueNumbers(constantsProperties.getGenerationBathSize());
+        List<String> newHashes = encoder.encode(uniqueNumbers);
+        hashRepository.save(newHashes);
```
давай в качестве обучения добавим кастомную метрику в случае если не получилось сохранить хэши в бд

miked4king commented on src/main/java/faang/school/urlshortenerservice/HashGenerator.java:
```diff
@@ -0,0 +1,35 @@
+        hashRepository.save(newHashes);
+    }
+
+    @PostConstruct
+    private void settingsInitializer() {
+        generatorThreshold =
+                Math.max(10L * constantsProperties.getGenerationThresholdPercent(),
+                        constantsProperties.getLocalCachingSize());
+    }
```
метод расположить после полей класса

miked4king commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepository.java:
```diff
@@ -0,0 +1,55 @@
+public class HashRepository {
+    private final JdbcTemplate jdbcTemplate;
+
+    @Value("${spring.constants.local_caching_size}")
```
spring - как будто не спринговый параметр
возможно добавить конкретики в пути

miked4king commented on src/main/java/faang/school/urlshortenerservice/controller/UrlController.java:
```diff
@@ -0,0 +1,30 @@
+    private final UrlService urlService;
+
+    @PostMapping("/url")
+    public String uploadLong(@RequestBody LongUrlDto longUrlDto) {
```
давай вернем dto

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public String findOriginal(String hash) {
```
давай сделаем название поинформативней

miked4king commented on src/main/java/faang/school/urlshortenerservice/service/UrlServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+                hash.chars().allMatch(c -> hashEncoderProperties.getAlphabet().indexOf(c) >= 0)) {
+            return;
+        }
+        throw new IllegalArgumentException("Provided hash wrong format");
```
тут и в 40 поинформативней сообщения

miked4king commented on src/main/java/faang/school/urlshortenerservice/CleanerScheduler.java:
```diff
@@ -0,0 +1,44 @@
+        return removedHashes.size();
+    }
+
+    private CleanerScheduler getSelf() {
+        return (CleanerScheduler) AopContext.currentProxy();
+    }
```
давай переработаем без этого.
cleanUpBatch будем в отдельном сервисе вызывать

miked4king commented on src/main/java/faang/school/urlshortenerservice/CleanerScheduler.java:
```diff
@@ -0,0 +1,44 @@
+    private final UrlRepository urlRepository;
+    private final ConstantsProperties constantsProperties;
+
+    @Scheduled(cron = "${spring.scheduling.expired_hash_cleanup}")
```
https://www.baeldung.com/shedlock-spring почитать для ee

miked4king commented on src/main/java/faang/school/urlshortenerservice/async/AsyncConfig.java:
```diff
@@ -0,0 +1,27 @@
+@RequiredArgsConstructor
+public class AsyncConfig {
+
+    private final AsyncProperties properties;
```
можно в аргументы метода передать если только в 1 месте используются

miked4king commented on src/main/java/faang/school/urlshortenerservice/config/ConstantsProperties.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.stereotype.Component;
+
+@Data
+@Component
```
попробуй на мэйн класс повесить @ConfigurationPropertiesScan, вроде не требуется после нее вешать @Component на конф классы

miked4king commented on src/main/java/faang/school/urlshortenerservice/exception/UrlExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+    @ExceptionHandler(EntityNotFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponseDto handleNotFound(EntityNotFoundException e) {
+        log.error("EntityNotFoundException with message {} was thrown", e.getMessage());
```
лучше логгировать как можно больше информации об ошибках.
например залоггировать ErrorResponseDto и полностью ошибку EntityNotFoundException

AlexandrSid committed:
url shortener.
After review refactoring
Clean up project root folder

AlexandrSid committed:
url shortener.
After review refactoring
Interfaces added to repositories

AlexandrSid committed:
url shortener.
After review refactoring
configs refactoring
PostConstruct order fixed

AlexandrSid committed:
url shortener.
After review refactoring
HashCache interface added,
Cache check and refill invoke logic changed
scheduling settings refactored

AlexandrSid committed:
url shortener.
After review refactoring
Exceptions handling and creating refactoring
More detailed messages, full stacktrace logging, api/v added to controller and filter

AlexandrSid committed:
url shortener.
After review refactoring
Renaming

AlexandrSid committed:
url shortener.
After review refactoring
Autoproxing removed, interface creation is discussible

AlexandrSid committed:
ARR (After review refactoring)
Base62Encoder logic changed.
Squids lib - new hash provider.
https://github.com/sqids/sqids-java
DB migrations merged.

AlexandrSid committed:
ARR (After review refactoring)
Base62Encoder logic changed.
Squids lib - new hash provider.
https://github.com/sqids/sqids-java
DB migrations merged.

AlexandrSid committed:
Merge remote-tracking branch 'origin/feature/BJS2-84622' into feature/BJS2-84622

AlexandrSid committed:
ARR (After review refactoring)
Base62Encoder logic changed.
Expired Urls cleaner now synchronised between instances

AlexandrSid committed:
ARR (After review refactoring)
Basic metrics collection added

AlexandrSid committed:
ARR (After review refactoring)
failedSavedCounter added. with fake increment logic

AlexandrSid committed:
ARR (After review refactoring)
real saved errors count now

AlexandrSid committed:
ARR (After review refactoring)
real saved errors count now

AlexandrSid committed:
Merge remote-tracking branch 'origin/feature/BJS2-84622' into feature/BJS2-84622

AlexandrSid committed:
After dockerisation and hi-load test
cache run low protection added to HashCache

AlexandrSid committed:
hash_cache_size metric added

miked4king commented on src/main/java/faang/school/urlshortenerservice/hash/CleanerScheduler.java:
```diff
@@ -0,0 +1,41 @@
+package faang.school.urlshortenerservice.hash;
```
давай вынесем в пакет scheduler в корне проекта

AlexandrSid replied:
Я переименовал класс в ExpiredUrlsHashCleanScheduler для лучшей реперезентации названием его логики.
В ситуации, когда в сервисе нет хотя бы 3х cron задач, которые уже следут конфигурировать централизованно, рекомедуется держать содержащие их классы в доменных папках. Это соответсвует принципам high cohesion и YAGNI.

miked4king commented on src/main/java/faang/school/urlshortenerservice/hash/HashGenerator.java:
```diff
@@ -0,0 +1,39 @@
+
+@Component
+@RequiredArgsConstructor
+public class HashGenerator {
```
интерфейс потеряли? вроде был

AlexandrSid replied:
не было. Хотя и другая реализация HashGenerator на замену этой возможна, такая, которая бы пополняла таблицу хэшей не опираясь на Base62 алгоритм и DB sequence, но до тех пор пока она не рассматривается даже в планах, вынесение интерфейса для HashGenerator является нарушением принципов DRY и YAGNI

miked4king commented on src/main/java/faang/school/urlshortenerservice/repository/HashRepositoryJdbcImpl.java:
```diff
@@ -0,0 +1,59 @@
+    private final JdbcTemplate jdbcTemplate;
+
+    @Value("${constants.local_hash_cache_butch_size}")
+    private int LOCAL_BATCH_SIZE;
```
константы же static final, нужно название переменной в camelCase сделать

miked4king commented on src/main/java/faang/school/urlshortenerservice/util/LockUtil.java:
```diff
@@ -0,0 +1,15 @@
+package faang.school.urlshortenerservice.util;
+
+import java.util.concurrent.locks.ReentrantLock;
+
```
можем добавить аннотацию @UtilityClass

AlexandrSid committed:
refactoring II
Value specified field replaced with property,
Class renamed for better logic representation,
UtilityClass annotation added
